### PR TITLE
oc: switch to OKAPI

### LINKED ISSUE #9: Switch to OKAPI for OC* services
See https://github.com/opencaching/opencaching-pl/blob/master/xml/README.txt and https://github.com/geokrety/geokrety-website/issues/427

GeoKrety should use OKAPI for fetching data from OC.pl.

- Check other websites of OC network, is there any okapi usage already in place?
- If yes, then it should be simple enough to adopt some of existing scripts for .pl. If no, need to develop some okapi-wrapper

### PR DESCRIPTION
**Pull request by okainov**
Use OKAPI instead of old out-of-date XML API

Fixes #9

To be merged after https://github.com/geokrety/geokrety-website/pull/429 and #10 are merged and applied.

### CODE REVIEW

**Comment by okainov** [comment]
OC.US have some issues with API sync - they provide `fulldump` which is older than min accessible revision from `changelog`. I contacted them via email, didn't find any github-like place.

Hopefully just a temporary glitch.

And OC.PL dump for some reason is not being unpacked via Phar, I got it working manually, but still a bit suspicious.

**Comment by okainov** [comment]
@kumy I was not really sure what "status" is, so I just skipped it =)

Also it was a bit confused to have both `kraj` and `country` in the DB, so I just used `kraj` to be consistent with original script... But this should be cleaned up.

**Comment by kumy** [comment]
Thanks for PRs. I will check all that nice work when back at home tomorrow.

**Comment by kumy** [comment]
Will check that in a moment, it's sad but we don't have staging environment for those scripts/repo :(

File: waypointy/oc/xml2sql.php [hunk 1]
```diff
@@ -1,166 +1,345 @@
...
-
-    //print_r($xml); die();
-
-    echo " * processing ". $baza['prefix'] ." ($key)\n";
-    echo " *      count:". $nbImported . "\n";
-    echo " *        url:". $baza['url'] . $modifiedsince . "\n";
-    if ($xml->cache) foreach ($xml->cache as $cache) {
+    $success = true;
+    if ($fullResync) {
+        $fullDumpPath = 'oc_dump_extracted';
```
Comment (kumy) [inline waypointy/oc/xml2sql.php:378]: Would have been nice if this directory could be moved out of source tree, as it may become readonly in the future.

idea: https://stackoverflow.com/a/30010928/944936

File: waypointy/oc/xml2sql.php [hunk 2]
```diff
@@ -1,166 +1,345 @@
...
+    $files = array_diff(scandir($dir), array('.', '..'));
+    foreach ($files as $file) {
+        (is_dir("$dir/$file")) ? deleteTree("$dir/$file") : unlink("$dir/$file");
+    }
+    rmdir($dir);
+}
+
+/**
+ *
+ * Copies the file from $url to $output, supports both file paths and urs
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: typo `urs` -> `url` (`uris`?)

File: waypointy/oc/xml2sql.php [hunk 3]
```diff
@@ -1,166 +1,345 @@
...
 
+        $id = $change->object_key->code;
 
-// $BAZY_OC['no']['prefix'] = 'OS';
-// $BAZY_OC['no']['url'] = "http://www.opencaching.no/xml/ocxml11.php?cache=1&session=0&charset=utf-8&cdata=1&xmldecl=1&ocxmltag=1&doctype=0&zip=0&modifiedsince=";
-// $BAZY_OC['no']['szukaj'] = 'http://www.opencaching.no/searchplugin.php?userinput=';
+        if ($change->change_type == 'delete') {
+            //delete from DB
 
+            $sql = 'DELETE FROM `gk-waypointy` WHERE waypoint="' . mysqli_real_escape_string($link, $id) . '"';
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: Would you please convert queries to prepared statements?

> Prepared statements also provide strong protection against SQL injection
https://www.tutorialrepublic.com/php-tutorial/php-mysql-prepared-statements.php

File: waypointy/oc/xml2sql.php [hunk 4]
```diff
@@ -1,166 +1,345 @@
...
+            // So we need to trigger actual update only if at least one of our fields was changes.
+
+            $sqlInsert [] = 'waypoint';
+            $sqlValues [] = "'$id'";
+            $sqlUpdate [] = "waypoint='$id'";
+
+            $insertPart = '(' . implode(',', $sqlInsert) . ')';
+            $valuesPart = '(' . implode(',', $sqlValues) . ')';
+            $onDupPart = implode(',', $sqlUpdate);
+            $sql = 'INSERT INTO `gk-waypointy` ' . $insertPart . ' VALUES ' . $valuesPart . ' ON DUPLICATE KEY UPDATE ' . $onDupPart;
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: prepared statement too.

File: waypointy/oc/xml2sql.php [hunk 5]
```diff
@@ -1,166 +1,345 @@
...
+            $sqlValues [] = "'$owner'";
+            $sqlUpdate [] = "owner='$owner'";
+        }
+        if (isset($change->data->location)) {
+            $location = explode('|', $change->data->location);
+            $lon = mysqli_real_escape_string($link, (string)$location[0]);
+            $lat = mysqli_real_escape_string($link, (string)$location[1]);
+
+            $sqlInsert [] = 'lon';
+            $sqlValues [] = "'$lon'";
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: Does float values need surround quotes? However, don't care as prepared statement is needed.

File: waypointy/oc/xml2sql.php [hunk 6]
```diff
@@ -1,166 +1,345 @@
...
+        }
+        if (isset($change->data->type)) {
+            $type = mysqli_real_escape_string($link, (string)$change->data->type);
+            $sqlInsert [] = 'typ';
+            $sqlValues [] = "'$type'";
+            $sqlUpdate [] = "typ='$type'";
+        }
+        if (isset($change->data->country)) {
+            $country = mysqli_real_escape_string($link, (string)$change->data->country);
+            $sqlInsert [] = 'kraj';
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: > Also it was a bit confused to have both kraj and country in the DB, so I just used kraj to be consistent with original script... But this should be cleaned up.

`kraj` is the country code as ISO 3166-1 alpha-2 (https://en.wikipedia.org/wiki/ISO_3166-1_alpha-2). `country` is the full English country name. `kraj` is used in `new-theme`, `country` will not. `country` was used in `ruchy-ajax` lookups on legacy implementation.

File: waypointy/oc/xml2sql.php [hunk 7]
```diff
@@ -1,166 +1,345 @@
...
+
+            if ($result === false) { // ooooPs we got an import error !
+                print("error sql : id:$id - query:$sql - mysqli_error:" . mysqli_error($link) . "\n");
+            }
+            $nUpdated++;
+        }
+    }
+
+    echo " *    updated:" . $nUpdated . "\n";
+    echo " *    deleted:" . $nDeleted . "\n";
```
Comment (kumy) [inline waypointy/oc/xml2sql.php:229]: Is there a way to have stats for `created` extracted as an individual value? I'm unsure how it can be differentiated within an `ON DUPLICATE` query.

File: waypointy/oc/xml2sql.php [hunk 8]
```diff
@@ -1,166 +1,345 @@
...
+        $changes = json_decode(file_get_contents($folder . '/' . $piece));
+        performIncrementalUpdate($link, $changes);
+    }
+    return $revision;
+}
+
+function getLastUpdate($service)
+{
+    $link = DBPConnect();
+    $sql = "SELECT `last_update` FROM `gk-waypointy-sync` WHERE `service_id` LIKE '" . $service . "%'";
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: Why do you use `LIKE` and not an equality comparator?

File: waypointy/oc/xml2sql.php [hunk 9]
```diff
@@ -1,166 +1,345 @@
...
+    foreach ($index->data_files as $piece) {
+        $changes = json_decode(file_get_contents($folder . '/' . $piece));
+        performIncrementalUpdate($link, $changes);
+    }
+    return $revision;
+}
+
+function getLastUpdate($service)
+{
+    $link = DBPConnect();
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: It would be better if you could switch to the latest implementation of db connection like: `$link = GKDB::getLink();` it has the advantage to hold the link as a singleton. 

https://github.com/geokrety/geokrety-website/blob/master/website/lib/GKDB.class.php

https://github.com/geokrety/geokrety-website/blob/752f435dfcafd3f85f42ab2e0ecaaff763f150de/website/szukaj-ajax.php#L87

It will require to include the Class using `composer autoload`. Ex: 

https://github.com/geokrety/geokrety-website/blob/752f435dfcafd3f85f42ab2e0ecaaff763f150de/website/__sentry.php#L19-L20

File: waypointy/oc/xml2sql.php [hunk 10]
```diff
@@ -1,166 +1,345 @@
...
+
+    // Save bz2 file
+    $bz = bzopen($temp, "r");
+    $res = fopen("tempfile.tar", "w");
+    $copied = stream_copy_to_stream($bz, $res);
+    fclose($res);
+    fclose($bz);
+
+    // Extract tar
+    //For at least .pl Phar complains that archive is corrupted :(
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: Was this issue reported upstream? It would be better if we don't have to support exception here.

File: waypointy/oc/xml2sql.php [hunk 11]
```diff
@@ -1,166 +1,345 @@
...
+    // Maybe use exec on lin?
+    //exec('mkdir TestBlable && tar -C TestBlable -xvf tempfile.tar');
+
+    $phar = new PharData('tempfile.tar');
+    $phar->extractTo($output); // extract all files
+
+    // Finally we got 2+ Gb of random json files. Let's parse
```
Comment (kumy) [inline waypointy/oc/xml2sql.php:300]: Will those 2Gb be downloaded at each sync (it happens every 10min!)?
EDIT: okay, only on full import.

**Comment by kumy** [comment]
https://github.com/geokrety/geokrety-website/pull/429 need to be merged and applied before running this.

**Review by kumy — CHANGES_REQUESTED** [review]
Looks really nice thanks :+1:.
Could you please convert all database queries to prepared statements and use GKlink class to connect to the database. `DBPConnect` is obsolete and will be removed one day.

File: waypointy/oc/xml2sql.php (hunk 7 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php:229]: Actually I was not really sure how this output is needed/used at all. In some version during development I just had all this output removed...

File: waypointy/oc/xml2sql.php (hunk 10 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: I'm not quite sure which upstream do you mean. 

If OC.PL - The archive itself seems okayish (I tried "Test archive" with desktop 7Zip, also I think Winrar extracted it without any issues, just tooo slooow). 
If Phar - no, not reported, either not sure that it's some real issue or just random glitch. If you could try this on your machine\server\staging\whatever we could at least compare our results.

File: waypointy/oc/xml2sql.php (hunk 6 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: I just put whatever data coming from OC website here :)

Final dataset looks somehow like this
image

File: waypointy/oc/xml2sql.php (hunk 5 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: Probably not, will fix.

Though I'm not really sure how to use prepared statement for this type of logic when I compose the whole query on the fly, when amount of fields depends on the input

File: waypointy/oc/xml2sql.php (hunk 3 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: Will try :)

File: waypointy/oc/xml2sql.php (hunk 2 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: done

File: waypointy/oc/xml2sql.php (hunk 8 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: done

File: waypointy/oc/xml2sql.php (hunk 3 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: done

File: waypointy/oc/xml2sql.php (hunk 5 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: done. However for especially this query I don't really like idea of using anonymous prepared statement (`?`) because it will make the code much less readable IMO. SO left the query as is

Comment (kumy) [reply waypointy/oc/xml2sql.php]: Security over readability please 🤲. We should not trust content even if it come from "partners".

Comment (kumy) [reply waypointy/oc/xml2sql.php]: https://stackoverflow.com/a/15786451/944936

File: waypointy/oc/xml2sql.php (hunk 4 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: done

File: waypointy/oc/xml2sql.php (hunk 9 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: Since _setry include was already here, then it should work out-of-the box

File: waypointy/oc/xml2sql.php (hunk 5 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php]: There is de-facto no difference between using prepared statement and just proper escaping preventing SQL injections here. Security is completely the same. But whatever, changed.

File: waypointy/oc/xml2sql.php (hunk 7 above)
Comment (okainov) [reply waypointy/oc/xml2sql.php:229]: If you don't mind one extra query per each update ("does this GK exist in the table?") then I can make it properly separated with two queries (`INSERT...` and `UPDATE...`)

File: waypointy/oc/xml2sql.php [hunk 12]
```diff
@@ -1,166 +1,350 @@
...
-*/
+        if (sizeof($sqlInsert) > 0) {
+            // It can happen that changelog contains useless fields like "founds" which we are not interested in.
+            // So we need to trigger actual update only if at least one of our fields was changes.
 
+            $sqlInsert [] = 'waypoint';
+            $sqlValues [] = $id;
+            $sqlUpdate [] = "waypoint=?";
 
-include_once("../../konfig-tools.php");
```
Comment (kumy) [inline waypointy/oc/xml2sql.php:50]: This include is necessary, as it define `$geokrety_www`

Comment (okainov) [reply waypointy/oc/xml2sql.php:50]: oops

**Comment by kumy** [comment]
First test raise error
```
Fatal error: Uncaught Error: Call to undefined function bzopen() in /opt/geokrety/waypointy/oc/xml2sql.php:214
Stack trace:
#0 /opt/geokrety/waypointy/oc/xml2sql.php(309): saveAndExtractDumpFile('https://opencac...', 'oc_dump_extract...')
#1 {main}
  thrown in /opt/geokrety/waypointy/oc/xml2sql.php on line 214
```
one extension seems absent in our installation.

(I have to leave, I'll check later ;))

File: waypointy/oc/xml2sql.php [hunk 13]
```diff
@@ -1,166 +1,351 @@
...
+        if (isset($change->data->owner->username)) {
+            $owner = mysqli_real_escape_string($link, (string)$change->data->owner->username);
+            $sqlInsert [] = 'owner';
+            $sqlValues [] = $owner;
+            $sqlUpdate [] = "owner=?";
+        }
+        if (isset($change->data->location)) {
+            $location = explode('|', $change->data->location);
+            $lon = floatval($location[0]);
+            $lat = floatval($location[1]);
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: Values are reversed…

File: waypointy/oc/xml2sql.php [hunk 14]
```diff
@@ -1,166 +1,409 @@
...
+function deleteTree($dir)
+{
+    if (empty($dir) or !$dir) {
+        return;
+    }
+    $pathLength = strlen($dir);
+    $files = array_diff(scandir($dir), array('.', '..'));
+    foreach ($files as $file) {
+        $file_ = realpath("$dir/$file");
+        if (strncmp($file_, $pathLength) !== 0) {
```
Comment (kumy) [inline waypointy/oc/xml2sql.php]: I must admit this part was not tested yet :/

File: waypointy/oc/xml2sql.php (hunk 10 above)
Comment (kumy) [reply waypointy/oc/xml2sql.php]: In my tests I was on php 7.3 and didn't get any error there.

**Comment by kumy** [comment]
@okainov I've added more commits, don't forget to rebase your local copy ;)

I've also added feature dependencies with https://github.com/geokrety/geokrety-website/pull/434, you'll have to pull them too.

File: waypointy/oc/xml2sql.php (hunk 7 above)
Comment (kumy) [reply waypointy/oc/xml2sql.php:229]: > "does this GK exist in the table?"

You mean: "does this **Waypoint** exist in the table?"

I'have introduced such request, your may use the result if you wish to implement it. Else let's forget about it, it's not that important.

File: waypointy/oc/xml2sql.php (hunk 1 above)
Comment (kumy) [reply waypointy/oc/xml2sql.php:378]: Still need to implement "working outside code tree"

**Comment by kumy** [comment]
Hum implementation will still need some tuning regarding prepared statements…

```
Fatal error: Uncaught Exception: insertOrUpdateWaypoint prepare failed: (1461) Can't create more than max_prepared_stmt_count statements (current value: 16382) in /var/www/html/lib/GKDB.class.php:76
Stack trace:
#0 /opt/geokrety/waypointy/oc/xml2sql.php(8): GKDB::prepareBindExecute('insertOrUpdateW...', 'INSERT INTO `gk...', 'ddsisddsis', Array)
#1 /opt/geokrety/waypointy/oc/xml2sql.php(227): prepareBindExecute('insertOrUpdateW...', 'INSERT INTO `gk...', 'ddsisddsis', Array)
#2 /opt/geokrety/waypointy/oc/xml2sql.php(243): performIncrementalUpdate(Array)
#3 /opt/geokrety/waypointy/oc/xml2sql.php(386): insertFromFullDump('oc_dump_extract...')
#4 {main}
  thrown in /var/www/html/lib/GKDB.class.php on line 76
```

~~EDIT: should be fixed using https://github.com/geokrety/geokrety-website/pull/435~~
EDIT2: The implementation was prone to memory leaks, so let's just close statements even if it seem less efficient…

**Comment by kumy** [comment]
Script is working is very nice, thanks @okainov :+1: 

I'll finish synchronizing db on my side, then import fresh waypoints into prod database and finally merge/deploy this PR.

File: waypointy/oc/xml2sql.php (hunk 1 above)
Comment (kumy) [reply waypointy/oc/xml2sql.php:378]: That could be a dedicated issue.

File: waypointy/oc/xml2sql.php (hunk 10 above)
Comment (kumy) [reply waypointy/oc/xml2sql.php]: I got the error with oc.ro, for the production deployment I've enabled the `exec()` version.

**Comment by kumy** [comment]
It's now on production :champagne:
